README.md
# php-google-drive

- [Installation](#installation)
- [Usage](#usage)
- [How to get Google Drive API credentials](#how-to-get-google-drive-api-credentials)
  - [Create a new project](#create-a-new-project)
  - [Enable Google Drive API](#enable-google-drive-api)
  - [Create credentials](#create-credentials)
  - [Publish app](#publish-app)
- [How to get a token](#how-to-get-a-token)

## Installation

``` sh
composer require w3lifer/php-google-drive
```

## Usage

``` php
<?php

// require_once __DIR__ . '/vendor/autoload.php';

use w3lifer\google\Drive;

$googleDrive = new Drive([
    'pathToCredentials' => __DIR__ . '/credentials.json', // Required
    'pathToToken' => __DIR__ . '/token.json', // Required
]);

$fileId = $googleDrive->upload(
    __DIR__ . '/hello.txt',  // Required
    [ // Optional
        '<folder id>',
        '<folder id>',
    ]
);
```

- Folder ID: `https://drive.google.com/drive/folders/<folder-id>`

## How to get Google Drive API credentials

### Create a new project

---

- https://console.cloud.google.com

> ![](https://i.stack.imgur.com/D0orV.png)

> ![](https://i.stack.imgur.com/dtqvu.png)

---

### Enable Google Drive API

---

> ![](https://i.stack.imgur.com/5S44Q.png)

> ![](https://i.stack.imgur.com/QEP5e.png)

> ![](https://i.stack.imgur.com/70YlU.png)

> ![](https://i.stack.imgur.com/Gz3gQ.png)

---

### Create credentials

---

> ![](https://i.stack.imgur.com/l9zBr.png)

<ins>If you left the screen above, go to "Enable API & services" and click "Google Drive API":</ins>

> ![](https://i.stack.imgur.com/jIjyu.png)

> ![](https://i.stack.imgur.com/BymkG.png)

> ![](https://i.stack.imgur.com/wkBrU.png)

> ![](https://i.stack.imgur.com/vBdC1.png)

> ![](https://i.stack.imgur.com/qXN5E.png)

> ![](https://i.stack.imgur.com/olM22.png)

---

### Publish app

---

> ![](https://i.stack.imgur.com/2txzO.png)

## How to get a token

**1. Save the credential to disk and specify the path to them by setting the `pathToCredentials` configuration key**

``` php
$googleDrive = new Drive([
    'pathToCredentials' => __DIR__ . '/credentials.json', // Required
    'pathToToken' => __DIR__ . '/token.json', // Required
]);
```

> **Note** that `pathToToken` is the path where the `Drive` saves the token after the first run. That is, the token will be saved automatically along the specified path; it is needed for your application to subsequently access Google Drive without the consent screen.

---

**2. Run your app**

- You will receive the following message in the console:

```
Open the following link in your browser:
https://accounts.google.com/o/oauth2/v2/auth?...
Enter verification code:
```

- Open the link in the message and choose desired account:

![](https://i.stack.imgur.com/lOzkM.png)

- The warning "Google hasn't verified this app" may appear. Click "Advanced" and then "Go to ... (unsafe)":

![](https://i.stack.imgur.com/YACgh.png)

- On the next screen click "Continue":

![](https://i.stack.imgur.com/wzWfm.png)

- After the redirect, copy the code from the address bar of your browser and paste it in the console:

![](https://i.stack.imgur.com/ZP45S.png)

```
Open the following link in your browser:
https://accounts.google.com/o/oauth2/v2/auth?...
Enter verification code: <here>
```

**3. Enjoy!**


CHANGELOG.md
# Change Log

- [3.0.0 November 17, 2023](#300-november-17-2023)
- [2.0.0 November 17, 2023](#200-november-17-2023)
- [1.1.0 January 6, 2020](#110-january-6-2020)
- [1.0.0 May 24, 2019](#100-may-24-2019)

## 3.0.0 November 17, 2023

- Main class renamed: `PhpGoogleDrive` → `Drive`

## 2.0.0 November 17, 2023

- Project renamed: `google-drive` → `php-google-drive`
- Main class renamed: `GoogleDrive` → `PhpGoogleDrive`

## 1.1.0 January 6, 2020

- Added `createFolder()` method

## 1.0.0 May 24, 2019

- Initial release


Drive.php
<?php

namespace w3lifer\google;

use Exception;
use Google_Client;
use Google_Service_Drive;
use Google_Service_Drive_DriveFile;

/**
 * @see https://github.com/googleapis/google-api-php-client/blob/master/examples/simple-file-upload.php
 */
class Drive
{
    private $pathToCredentials;

    private $pathToToken;

    private $client;

    private $service;

    public function __construct($config)
    {
        if (
            !empty($config['pathToCredentials']) &&
            file_exists($config['pathToCredentials'])
        ) {
            $this->pathToCredentials = $config['pathToCredentials'];
        } else {
            throw new Exception('Incorrect path to credentials');
        }

        if (empty($config['pathToToken'])) {
            throw new Exception('The path to token can not be empty');
        }
        $this->pathToToken = $config['pathToToken'];

        $this->client = new Google_Client();

        $this->client->setApplicationName('');
        $this->client->setAccessType('offline');

        $this->client->setAuthConfig($this->pathToCredentials);
        $this->client->addScope('https://www.googleapis.com/auth/drive');

        $this->checkAccessToken();

        $this->service = new Google_Service_Drive($this->client);
    }

    private function checkAccessToken()
    {
        if (file_exists($this->pathToToken)) {
            $accessToken = json_decode(
                file_get_contents($this->pathToToken),
                true
            );
            $this->client->setAccessToken($accessToken);
        }
        if ($this->client->isAccessTokenExpired()) {
            if ($this->client->getRefreshToken()) {
                $this->client->fetchAccessTokenWithRefreshToken($this->client->getRefreshToken());
            } else {
                $authUrl = $this->client->createAuthUrl();
                echo 'Open the following link in your browser:' . "\n" . $authUrl . "\n";
                echo 'Enter verification code: ';
                $authCode = trim(fgets(STDIN));
                $accessToken = $this->client->fetchAccessTokenWithAuthCode($authCode);
                $this->client->setAccessToken($accessToken);
                if (array_key_exists('error', $accessToken)) {
                    throw new Exception(join(', ', $accessToken));
                }
            }
            if (!file_exists(dirname($this->pathToToken))) {
                mkdir(dirname($this->pathToToken), 0700, true);
            }
            file_put_contents($this->pathToToken, json_encode($this->client->getAccessToken()));
        }
    }

    public function upload(string $pathToFile, array $folderIds = [])
    {
        $fileMetadata = new Google_Service_Drive_DriveFile([
            'name' => basename($pathToFile),
            'parents' => $folderIds,
        ]);
        $result = $this->service->files->create(
            $fileMetadata,
            [
                'data' => file_get_contents($pathToFile),
                'mimeType' => 'application/octet-stream',
                'uploadType' => 'multipart',
            ]
        );
        return $result->id;
    }

    /**
     * @see https://stackoverflow.com/a/58611113/4223982
     */
    public function createFolder(string $name)
    {
        $file = new Google_Service_Drive_DriveFile();
        $file->setName($name);
        $file->setMimeType('application/vnd.google-apps.folder');
        $result = $this->service->files->create($file);
        return $result->id;
    }
}


PhpGoogleDrive.php
<?php

namespace w3lifer\google;

use Exception;
use Google_Client;
use Google_Service_Drive;
use Google_Service_Drive_DriveFile;

/**
 * @see https://github.com/googleapis/google-api-php-client/blob/master/examples/simple-file-upload.php
 */
class PhpGoogleDrive
{
    private $pathToCredentials;

    private $pathToToken;

    private $client;

    private $service;

    public function __construct($config)
    {
        if (
            !empty($config['pathToCredentials']) &&
            file_exists($config['pathToCredentials'])
        ) {
            $this->pathToCredentials = $config['pathToCredentials'];
        } else {
            throw new Exception('Incorrect path to credentials');
        }

        if (empty($config['pathToToken'])) {
            throw new Exception('The path to token can not be empty');
        }
        $this->pathToToken = $config['pathToToken'];

        $this->client = new Google_Client();

        $this->client->setApplicationName('');
        $this->client->setAccessType('offline');

        $this->client->setAuthConfig($this->pathToCredentials);
        $this->client->addScope('https://www.googleapis.com/auth/drive');

        $this->checkAccessToken();

        $this->service = new Google_Service_Drive($this->client);
    }

    private function checkAccessToken()
    {
        if (file_exists($this->pathToToken)) {
            $accessToken = json_decode(
                file_get_contents($this->pathToToken),
                true
            );
            $this->client->setAccessToken($accessToken);
        }
        if ($this->client->isAccessTokenExpired()) {
            if ($this->client->getRefreshToken()) {
                $this->client->fetchAccessTokenWithRefreshToken($this->client->getRefreshToken());
            } else {
                $authUrl = $this->client->createAuthUrl();
                echo 'Open the following link in your browser:' . "\n" . $authUrl . "\n";
                echo 'Enter verification code: ';
                $authCode = trim(fgets(STDIN));
                $accessToken = $this->client->fetchAccessTokenWithAuthCode($authCode);
                $this->client->setAccessToken($accessToken);
                if (array_key_exists('error', $accessToken)) {
                    throw new Exception(join(', ', $accessToken));
                }
            }
            if (!file_exists(dirname($this->pathToToken))) {
                mkdir(dirname($this->pathToToken), 0700, true);
            }
            file_put_contents($this->pathToToken, json_encode($this->client->getAccessToken()));
        }
    }

    public function upload(string $pathToFile, array $folderIds = [])
    {
        $fileMetadata = new Google_Service_Drive_DriveFile([
            'name' => basename($pathToFile),
            'parents' => $folderIds,
        ]);
        $result = $this->service->files->create(
            $fileMetadata,
            [
                'data' => file_get_contents($pathToFile),
                'mimeType' => 'application/octet-stream',
                'uploadType' => 'multipart',
            ]
        );
        return $result->id;
    }

    /**
     * @see https://stackoverflow.com/a/58611113/4223982
     */
    public function createFolder(string $name)
    {
        $file = new Google_Service_Drive_DriveFile();
        $file->setName($name);
        $file->setMimeType('application/vnd.google-apps.folder');
        $result = $this->service->files->create($file);
        return $result->id;
    }
}
